docs/parser_plugins.md

```diff
@@ -1 +1,423 @@
+
+#### Mainfile matching
+
+Since mainfiles are intrinsically connected to the scripts that parse them, the directives should be set via `ParserEntryPoint`.
+I rare cases, however, an Oasis admin may decide to override these selection directives via the `nomad.yaml`.
+We publish our VASP XML parser in `src/nomad_parser_vasp/parsers/__init__.py` as
+
+```python
+from nomad.config.models.plugins import ParserEntryPoint
+
+class VasprunXMLEntryPoint(ParserEntryPoint):
+    def load(self):
+        from nomad_parser_vasp.parsers.xml_parser import VasprunXMLParser
+        return VasprunXMLParser(**self.dict())
+
+xml_entry_point = VasprunXMLEntryPoint(
+    name='VasprunXML Parser',
+    description='Parser for VASP output in XML format.',
+    mainfile_name_re='.*vasprun\.xml.*',
+)
+```
+
+`load` comes from the entry point system and should just return our parser (see below).
+The entry point itself specifies _parser data_ and the directives.
+There are three kinds of file aspects that can be targeted, all via _regular expressions_ (regex):
+
+- `mainfile_name_re` - the filename itself.
+- `mainfile_contents_re`, `mainfile_contents_dict` - the file contents. It is by default restricted to the first 1024 bytes, i.e. the file header.
+- `mainfile_mime_re` - the file mime.  <!-- TODO define file mime -->
+
+!!! abstract "Assignment"
+    XML is a common file extension, and the user may remove `vasprun` from the name.
+    Swap out `mainfile_name_re` for a different selection directive.
+    <!-- TODO research how multiple directives combine>
+
+!!! success "Solution"
+    VASP XML typically starts with the tags
+    ```xml
+    <?xml version="1.0" encoding="ISO-8859-1"?>
+    <modeling>
+    ```
+    You can capture this via `mainfile_contents_re` in a regex like `r'<\?xml version="1\.0" encoding="ISO\-8859\-1"?>\n<modeling>`. <!-- TODO double-check>
+
+#### Mainfile Interfacing
+
+Within the cascade, `MatchingParser`, acts as the connection point on the parser side.
+It plays less of a role in manipulating the directives, and more so in defining the _interface_ -a formalization of behavior- back to the parser.
+The two main specifications are instantiation and `parse`.
+Since `MatchingParser` already defines both, parsers may simply _inherit_ therefrom.
+The most rudimentary parser, thus looks as follows.
+
+```python
+from nomad.parsing import MatchingParser
+class VasprunXMLParser(MatchingParser):
+    pass
+```
+
+NOMAD can already run this parser, but will raise a `NotImplementedError`.
+The interface may be defined, but we still need to fill in the actual parsing by overwriting `parse`.
+
+!!! info "Run your parser"
+    In everyday NOMAD use, the user only interacts with NOMAD via the GUI or API.
+    NOMAD will regulate parsing as the user uploads via any of these channels.
+    During development, the command line is probably the preferable option, as you can load changes faster and incorporate it into your favorite test setups.
```

Comment:
Okay, I hadn't consider this way of working yet.
I think this would go better under an "Extra" section at the bottom, or an admonishment, as it breaks up the flow of writing a parser.